### PR TITLE
Editorial: Drop own powerful feature, require active local video source.

### LINKED ISSUE #79: Add camera permission requirement to spec?
See https://www.w3.org/2021/10/29-dap-minutes.html#t07:

> RESOLUTION: Add camera permission requirement to ALS spec to help enable shipment of the API

If I've understood it correctly, the idea would be to add text to the spec that mandates that permission for accessing ALS would be granted only if access to ALS as well as camera is allowed (in a similar fashion to `getUserMedia()` and the microphone/camera permissions, for example).

I'd like to make sure that this is indeed the idea, especially in light of https://github.com/w3c/ambient-light/issues/64#issuecomment-954444294 (_This will require continuous monitoring of the ALS and not make sense to bundle with a camera API (including permissions)_).

I'm also wondering if this makes sense as a normative requirement or as an implementation detail. The DAS F2F discussion mentioned Chrome's position:

> we'd be willing to enable the ALS API for websites that use camera already
> we could consider ALS augmenting the camera

in this case, isn't it up to the UA to do one of the following:
- Reject calls to `AmbientLightSensor.start()` if access to the camera has not been granted before
- Show a custom prompt for ALS that would cause access to camera and ALS to be granted together
- Automatically grant access to ALS if access to the camera has already been granted
- Something else entirely

### PR DESCRIPTION
**Pull request by rakuco**
Fixes #79.

The idea is that, based on the research on potential attacks on the Ambient Light Sensor API, it is important to prompt users before allowing access to illuminance readouts. This was already mandated by the main Generic Sensors spec, as `Sensor.start()` runs the "Request sensor access" abstract operation.

The challenge with the Ambient Light Sensor API is prompting users in a way that they understand what they are being prompted for; the assumption is that most users are not familiar with what an Ambient Light sensor is.

We have chosen to solve this issue by dropping our "ambient-light-sensor" powerful feature name altogether and integrating with the Media Capture and Streams specification instead: we consider an Ambient Light Sensor to be a 1x1 camera and require there to be at least one local video source that is not muted or stopped in order for illuminance readouts to be provided.

Per the Media Capture and Streams specification, this is only possible if script has called `MediaDevices.getUserMedia()` and granted the "camera" permission. This also means the User Agent has at least indicated to the user that a local video source has started being used.

In other words, an Ambient Light Sensor only provides readings if a local video source (such as a camera) is currently active and being used in the same window as the AmbientLightSensor instance, and when all local video sources stop we also stop providing readouts and fire an "error" with a NotReadableError exception.

### CODE REVIEW

**Comment by rakuco** [comment]
@reillyeon @anssiko please take a look at this first draft.

The Media Capture references are a bit hand-wavy because the spec itself is not very precise about its definitions of source and device or the way notifications about changes to the `[[devicesLiveMap]]` slot are supposed to be delivered.

Tying an AmbientLightSensor instance to a specific MediaStreamTrack would allow us to be a bit more precise in how we track the lifetime of the resources we need at the expense of making the API slightly more cumbersome.

I'd appreciate it if you can CC one or more people with more expertise in the Media Capture area to see if this approach makes sense.

**Comment by rakuco** [comment]
Also cc @willmorgan

File: index.bs [hunk 1]
```diff
@@ -283,6 +293,39 @@ experience. <a href="https://crbug.com/1332536">Chromium bug 1332536</a> and <a
...
+being aware that the data is being read.
+
+[[STEALINGSENSITIVEDATA]] and [[ALSPRIVACYANALYSIS]] specifically recommend
+requesting user permission before allowing access to illuminance readouts as a
+privacy measure. On the other hand, it can be argued that conveying to users
+what an Ambient Light Sensor is so that they can make an informed choice to
+grant or deny access to it can be difficult.
+
+What this specification does instead is consider an Ambient Light Sensor to be
+a 1x1 camera, integrate with the [[MEDIACAPTURE-STREAMS]] specification and
```
[the remark this answers is not in the document]
Comment (rakuco) [reply index.bs]: Fixed, thanks!

**Comment by rakuco** [comment]
cc @lknik too

**Comment by lknik** [comment]
So what's the mechanics of "require there to be at least one local video source that is not muted or stopped"? From as user's point of view is the query overtly about the camera?

**Comment by rakuco** [comment]
> So what's the mechanics of "require there to be at least one local video source that is not muted or stopped"? From as user's point of view is the query overtly about the camera?

From a user's point of view the idea is that this represents a camera, or something akin to it that also causes browsers to show an indicator to users that video capture is happening and that fulfills the `{video: true}` constraint when it is passed to `getUserMedia()`.

"source" is defined [here](https://w3c.github.io/mediacapture-main/#dfn-source), "local video" is a term I am not sure about to be honest.

Another way to look at it is that ALS readouts should be provided only when UAs also show users that video capture is happening according to https://w3c.github.io/mediacapture-main/#privacy-indicator-requirements -- I am still not sure if this spec should look at the `[[devicesLiveMap]]` internal slot or `anyVideoLive`.

File: index.bs [hunk 2]
```diff
@@ -283,6 +292,39 @@ experience. <a href="https://crbug.com/1332536">Chromium bug 1332536</a> and <a
...
+
+Many of the attacks on Ambient Light sensors referenced above rely on being
+able to access illuminance readouts for a certain amount of time without a user
+being aware that the data is being read.
+
+[[STEALINGSENSITIVEDATA]] and [[ALSPRIVACYANALYSIS]] specifically recommend
+requesting user permission before allowing access to illuminance readouts as a
+privacy measure. On the other hand, it can be argued that conveying to users
+what an Ambient Light Sensor is so that they can make an informed choice to
+grant or deny access to it can be difficult.
```
Comment (reillyeon) [inline index.bs:306]:
```suggestion
privacy measure. On the other hand, it can be difficult to convey to users
what an Ambient Light Sensor is so that they can make an informed choice to
grant or deny access to it.
```

File: index.bs [hunk 3]
```diff
@@ -283,6 +292,39 @@ experience. <a href="https://crbug.com/1332536">Chromium bug 1332536</a> and <a
...
+
+What this specification does instead is consider an Ambient Light Sensor to be
+a 1x1 grayscale camera, integrate with the [[MEDIACAPTURE-STREAMS]]
+specification and require there to be at least one local video [=source=] that
+is not [=muted=] or [=stopped=] in order for illuminance readouts to be
+provided. In other words, an Ambient Light Sensor only provides readings if a
+local video source (such as a camera) is currently active and being used in the
```
Comment (reillyeon) [inline index.bs:313]: Shouldn't we require that the video source be a camera? A screen capture session should qualify.

**Review by reillyeon — CHANGES_REQUESTED** [review]

File: index.bs (hunk 3 above)
Comment (rakuco) [reply index.bs]: Done, please take another look.

I've replaced most occurrences of "local video source" with "local camera source". I thought a screen capture session wouldn't count as a "local video source", but neither the Screen Capture nor the Media Capture specs seem to be very precise when referring to different source types.

I'm not entirely sure "local camera" is redundant though.

**Review by reillyeon — APPROVED** [review]

**Comment by rakuco** [comment]
Thanks for the reviews. I'd be glad if you have contacts who work on Media Capture who could help check these changes and suggest the best wording here.

**Comment by noamr** [comment]
Some drive-by comments (I'm not a reviewer to either of these specs, but I ran into it at the W3C permissions workshop):

- Usually I read "editorial" as changing wording without changing meaning/behavior. This seems like a big change in behavior. Are we using the word differently in different specs?
- The [use-cases](https://w3c.github.io/ambient-light/#usecases-requirements) in the spec contain some entries that this would not work with - smart home, lighting at the work place... Seems like this is about changing the purpose of the ALS API to be more of an API that supplements video captures (the 3rd and 4th use cases). So I would expect that this change modifies the use cases, and focuses on them in the examples.
- I don't think we should treat the ALS as a 1x1 camera. Because:
   1. It feels wrong (to me, though others at the permissions workshop indicated a similar sentiment).
   1. Do users know where that 1x1 camera is on their phone/laptop? I don't. I can cover or turn my camera away but I have no idea how to physically do that with my ambient light sensor. So if I start a video stream and then physically cover my camera for part of it, the stream would still include ambient light readings, which is probably not what I intended. 
   1. Purpose: cameras measure what you show them, ALS is ambient by definition, meaning it is meant to capture light without user involvement.
   
Suggestion: If the idea here is that you'd get an ambient-light stream together with your video stream (which is kind of how I understand this change, with the only example provided), perhaps a more direct way to do it is that AL would be an additional [MediaStreamTrack](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack) rather than a bespoke JS API? I think this still has some privacy/permission issues, but at least it would be more purpose-driven than capability-driven.

**Comment by alvestrand** [comment]
Seems somewhat scary - this (treating the light sensor as an 1x1 camera) would mean that the red "I am capturing you" light would light up on all tabs that access the ambient light sensor, which is usually a fairly strong indication of privacy exposure. We risk numbing the user to the red light indicator "because it's so frequently on".

**Comment by alvestrand** [comment]
We do have a number of features (IP address exposure, device enumeration) that have their permissions tied back to the camera permission, without being a camera. Copying that kind of access is a possibility.

**Comment by noamr** [comment]
> We do have a number of features (IP address exposure, device enumeration) that have their permissions tied back to the camera permission, without being a camera. Copying that kind of access is a possibility.

Device enumeration is tied to the camera only when you try to enumerate the camera, no? https://w3c.github.io/mediacapture-main/#camera-information-can-be-exposed

Which API is "IP address exposure"?

**Comment by rakuco** [comment]
> Seems somewhat scary - this (treating the light sensor as an 1x1 camera) would mean that the red "I am capturing you" light would light up on all tabs that access the ambient light sensor, which is usually a fairly strong indication of privacy exposure. We risk numbing the user to the red light indicator "because it's so frequently on".

"Treating the light sensor as a 1x1 camera" was used in the sense of "we want the privacy requirements to be as strict as if an actual camera was being used".

However, this spec does not _request_ permission, but only _requires_ it. In other words, it cannot turn on the red "I am capturing you" light on its own -- attempting to start the sensor without having camera capture already happening will result in an error. In this example, the red indicator would light up on multiple tabs if and only if all these tabs are also capturing video.

**Comment by rakuco** [comment]
Hey @noamr, thanks for the comments. The more feedback we get about this change, the better.

> Some drive-by comments (I'm not a reviewer to either of these specs, but I ran into it at the W3C permissions workshop):
> 
> * Usually I read "editorial" as changing wording without changing meaning/behavior. This seems like a big change in behavior. Are we using the word differently in different specs?

My bad. Looking at https://www.w3.org/2021/Process-20211102/#correction-classes again it does look like "editorial" does not apply here. I've updated the PR title and the relevant commit messages.

> [...]
>
> Suggestion: If the idea here is that you'd get an ambient-light stream together with your video stream (which is kind of how I understand this change, with the only example provided), perhaps a more direct way to do it is that AL would be an additional [MediaStreamTrack](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack) rather than a bespoke JS API? I think this still has some privacy/permission issues, but at least it would be more purpose-driven than capability-driven.

The idea of integrating ALS into getUserMedia() was discussed in https://github.com/w3c/ambient-light/issues/64#issuecomment-817129135 and https://github.com/w3c/ambient-light/issues/79#issuecomment-1175960515.

At the very least, this would require much bigger changes in the (current) implementation, as `//services/device/generic_sensor` and the Media Capture stack in Chromium do not communicate at all. There's also the question of whether data should be provided as a stream or only when a certain threshold is reached. Finally, if we define a new MediaStreamTrack type for ALS and require developers to do something like `getUserMedia({video: true, als: true})`, don't you end up in the same situation where you need to figure out a way to prompt users for access and explain to them what an ambient light sensor is?

You raise some interesting points about why an ALS shouldn't be treated like a camera, but I'd like to understand if going down the getUserMedia() route really makes sense or not.

**Comment by lknik** [comment]
> > So what's the mechanics of "require there to be at least one local video source that is not muted or stopped"? From as user's point of view is the query overtly about the camera?
> 
> From a user's point of view the idea is that this represents a camera, or something akin to it that also causes browsers to show an indicator to users that video capture is happening and that fulfills the `{video: true}` constraint when it is passed to `getUserMedia()`.

OK. In other words - the webpage is capturing the video, anyway, is that right? There may be a some constraints as 'where the camera is pointing and whether it's the same direction as ALS', but it's probably not necessary to go in such details.

**Comment by rakuco** [comment]
> OK. In other words - the webpage is capturing the video, anyway, is that right?

Correct. Illuminance data is reported if and only if the web page is capturing video from a camera.

**Comment by rakuco** [comment]
I've pushed a few changes to the Use Cases and Examples sections to remove the bits that do not make sense with the new camera requirement. The Use Cases list got quite small, so I'd appreciate it if people (like @willmorgan for example :-) could help think of a few more ideas to add there.

This addresses most of @noamr's points. IMO the remaining ones are "users do not know where their ambient light sensor is located" and "expose illuminance as a new MediaStreamTrack". I feel like doing the latter wouldn't help with the former. I couldn't find any examples of new MediaStreamTrack types that are not video or audio (unless I misunderstood the suggestion), and it is still unclear how to deal with prompting users if we do that.

**Comment by anssiko** [comment]
>I will report back when the proceedings from that workshop are published.

Published: https://www.w3.org/Privacy/permissions-ws-2022/report
